File: Unique Paths II/Unique Paths II.cpp
class Solution {
public:
    int uniquePathsWithObstacles(vector<vector<int> > &obstacleGrid) {
        
        std::vector<std::vector<int> >  record;
        int rowSize = obstacleGrid.size();
        if(!rowSize)                    return 0;
        int colSize = obstacleGrid[0].size();
        if(!colSize)                    return 0;
        
        for(unsigned int row = 0; row < rowSize; ++row){
            std::vector<int>    rowElement;
            for(unsigned int col = 0; col < colSize; ++col){
                rowElement.push_back(0);
            }
            record.push_back(rowElement);
        }
        
        for(unsigned int row = 0; row < rowSize; ++row){
            for(unsigned int col = 0; col < colSize; ++col){
                if(!row && !col){
                    record[row][col] = (!obstacleGrid[row][col]);
                }
                else if(!row){
                    record[row][col] = (record[row][col - 1] && !obstacleGrid[row][col]);
                }
                else if(!col){
                    record[row][col] = (record[row - 1][col] && !obstacleGrid[row][col]);
                }
                else{
                    if(!obstacleGrid[row][col]){
                        record[row][col] = record[row][col - 1] + record[row - 1][col];
                    }
                    else{
                        record[row][col] = 0;
                    }
                }
            }
        }
        
        return record[rowSize - 1][colSize - 1];
    }
};

File: Unique Paths II/Unique Paths II.java
public class Solution {
    public int uniquePathsWithObstacles(int[][] obstacleGrid) {
        if(obstacleGrid.length == 0 || obstacleGrid[0].length == 0)         return 0;
        int[][] dp = new int[obstacleGrid.length][obstacleGrid[0].length];
        
        for(int i = 0; i < obstacleGrid.length; ++i){
            for(int j = 0; j < obstacleGrid[0].length; ++j){
                if(i == 0 && j == 0){
                    if(obstacleGrid[i][j] == 1)     dp[i][j] = 0;
                    else                            dp[i][j] = 1;
                }
                else if(i == 0){
                    if(obstacleGrid[i][j] == 1)     dp[i][j] = 0;
                    else                            dp[i][j] = dp[i][j - 1];
                }
                else if(j == 0){
                    if(obstacleGrid[i][j] == 1)     dp[i][j] = 0;
                    else                            dp[i][j] = dp[i - 1][j];
                }
                else{
                    if(obstacleGrid[i][j] == 1)     dp[i][j] = 0;
                    else                            dp[i][j] = dp[i - 1][j] + dp[i][j - 1];
                }
            }
        }
        
        return dp[obstacleGrid.length - 1][obstacleGrid[0].length - 1];
    }
}
